src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -520,7 +519,7 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         }
         def checkOverrideTerm() {
           other.cookJavaRawInfo() // #2454
-          if (!overridesTypeInPrefix(lowType, highType, rootType)) { // 8
+          if (!overridesTypeInPrefix(lowType, highType, rootType, low.isModuleOrModuleClass || high.isModuleOrModuleClass)) { // 8
```

Comment:
Only in the special case where a module type is replaced by another (by overriding an object with another object with a conforming type) does the subtyping exception make sense. You couldn't replace a val with a singleton type by an object because the object would not also replace the singleton type.

Overriding an object by a val is possible but doesn't require any subtyping exception:

```
scala> class Foo { object X { def y: Int = 42 }; def x = X }
defined class Foo

scala> class Bar extends Foo { override val X: super.X.type = x }
defined class Bar

scala> (new Bar).X
res1: Bar#X.type = null
```

Since https://github.com/scala/scala/commit/f32a32b1b33c9d7ccd62467e3e10cb69930023c8 requires overriding vals to be non-lazy, this doesn't look very useful. At first glance, also allowing lazy vals to override objects may be sound but wouldn't add much utility.

I'll add some more test cases for these situations.